extensions/signature.md

```diff
@@ -0,0 +1,108 @@
+# Signature Extension
+
+This extension defines a standard mechanism to carry signatures allowing
+event consumers to ensure the identity of the event producer and verify
+the event-data has not been tampered with during transport.
+
+This specification does not:
+
+- Declare a mechanism for public key exchange.
+- Declare a mechanism for public key selection during signature verification.
+Implementors may consider leveraging the CloudEvent `source` as a method
+of selection.
+
+## Usage
+
+This extension supports a few general usage models:
+
+- Where the algorithm used to produce an attached signature needs to be shared.
+- Where the `data` is signed with a detached signature.
+- Where the CloudEvent context attributes are signed.
+
+## Attributes
+
+### sigdata
+
+- Type: `Binary`
+- Description: The binary signature generated for the event `data`.
+- Constraints:
+
+  - OPTIONAL
+
+### sigdataalgo
```

Comment:
Does this provide an ID of the materials allowing verification (e.g. certificate common name, other type of public key ID, or even a shared key ID)?

The signature will need to be decrypted with the right material(s) to verify possession of identity by the sender.  How are those selected?

Reply:
Good point .... there's a bit of a hand-wavy disclaimer at the begging around the selection of keys.

I'm open to suggestions ;-)

Reply:
Yes, I appreciate that you did.  The challenge is that keys roll over time.  While one could try and use a hopefully synchronized wall clock as a proxy for key roll over alignments over time but providing an attribute to write an identifier into is way easier if also duplicative across events.  So while I think it's right for the specification to stay out of the business of declaring the nature of the materials, facilitating the declaration of their identities (optionally if it makes sense) seems reasonable.

On a wider scope, I've made a suggestion on the top line comment of this review: that the signatures be an ordered array identifying the signed content, the resulting signature, the algorithm used, the id of the cryptographic materials used, and any altered values that were previously signed long with their prior values.  The algorithm for verification would take the event as given, verify it against the last signature of the event, modify the event to revert any alterations identified in the last signature, and pass the altered/reverted event along with the signature array (less the just verified signature) back into the same algorithm.  This process would exited with failure if any signature verification fails (or potentially only a subset of signatures that the receiver "cares about") but otherwise, once no more signatures exist to verify then verification is successful.  I apologize for offering what amounts to a bit of a rewrite of what you've offered but it seems like it might support a wider array of our scenarios.  Happy to own that work if you would prefer.